### PR TITLE
Fixed SOURCE_IP session persistence

### PR DESCRIPTION
**Pull request by mdurrant-b3**
* Corrected sessionpersistence <-> session_persistence mis-transposition
* Updated tests to reflect correct data structure

### CODE REVIEW

File: a10_neutron_lbaas/tests/unit/v2/fake_objs.py [hunk 1]
```diff
@@ -97,9 +97,9 @@ def __init__(self, protocol, method, persistence, listener=False,
         self.protocol = protocol
         self.lb_algorithm = method
         if persistence is None:
-            self.sessionpersistence = None
+            self.session_persistence = None
         else:
-            self.sessionpersistence = FakePersistence(persistence)
+            self.session_persistence = FakePersistence(persistence)
```
Comment (Cedev) [inline a10_neutron_lbaas/tests/unit/v2/fake_objs.py]: The tests should test it both ways for backwards/forwards compatibility.

File: a10_neutron_lbaas/v2/handler_persist.py [hunk 2]
```diff
@@ -40,8 +40,8 @@ def __init__(self, c, context, pool, old_pool=None, deprecated_arg=None):
         if pool:
             self.name = pool.id
 
-        if pool and pool.sessionpersistence:
-            self.sp = pool.sessionpersistence
+        if pool and pool.session_persistence:
+            self.sp = pool.session_persistence
```
Comment (Cedev) [inline a10_neutron_lbaas/v2/handler_persist.py]: It should work both ways for backwards/forwards compatibility.

File: a10_neutron_lbaas/v2/handler_pool.py [hunk 3]
```diff
@@ -88,22 +88,22 @@ def delete(self, context, pool):
 
     def _update_session_persistence(self, old_pool, pool, c, context):
         # didn't exist, does exist, create
-        if not old_pool or (not old_pool.sessionpersistence and pool.sessionpersistence):
+        if not old_pool or (not old_pool.session_persistence and pool.session_persistence):
```
Comment (Cedev) [inline a10_neutron_lbaas/v2/handler_pool.py]: It should work both ways for backwards/forwards compatibility.

**Comment by mdurrant-b3** [comment]
check

**Comment by Cedev** [comment]
I tried creating a loadbalancer, listener, and pool against a 2.7.2 vthunder

```
$ neutron lbaas-loadbalancer-create membersubnet --name vip001
Created a new loadbalancer:
+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| description         |                                      |
| id                  | a6f4c17d-351c-4f0a-a364-7f896a7b421d |
| listeners           |                                      |
| name                | vip001                               |
| operating_status    | ONLINE                               |
| provider            | a10networks                          |
| provisioning_status | ACTIVE                               |
| tenant_id           | e0142dcdf4904e05a40c2675db637473     |
| vip_address         | 10.90.102.187                        |
| vip_port_id         | ea7fdae9-ac5a-428e-81a1-be76855e9643 |
| vip_subnet_id       | a1188f60-75f9-4ce2-8a0e-94188f216343 |
+---------------------+--------------------------------------+
```

```
$ neutron lbaas-listener-create --name listener001 --loadbalancer vip001 --protocol HTTP --protocol-port 80
Created a new listener:
+---------------------------+------------------------------------------------+
| Field                     | Value                                          |
+---------------------------+------------------------------------------------+
| admin_state_up            | True                                           |
| connection_limit          | -1                                             |
| default_pool_id           |                                                |
| default_tls_container_ref |                                                |
| description               |                                                |
| id                        | 6df625e0-6fb7-4c1c-b109-81e9412c428a           |
| loadbalancers             | {"id": "a6f4c17d-351c-4f0a-a364-7f896a7b421d"} |
| name                      | listener001                                    |
| protocol                  | HTTP                                           |
| protocol_port             | 80                                             |
| sni_container_refs        |                                                |
| tenant_id                 | e0142dcdf4904e05a40c2675db637473               |
+---------------------------+------------------------------------------------+
```

```
s$ neutron lbaas-pool-create --name pool001 --lb-algorithm SOURCE_IP --listener listener001 --protocol HTTP --session-persistence type=SOURCE_IP
An error happened in the driver
```

I got the following error:

```
2017-02-08 16:02:04.367 ERROR neutron_lbaas.services.loadbalancer.plugin [req-6885af32-bfe7-4e5a-925f-f05a780d8338 admin e0142dcdf4904e05a40c2675db637473] There was an error in the driver
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin Traceback (most recent call last):
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/opt/stack/neutron-lbaas/neutron_lbaas/services/loadbalancer/plugin.py", line 471, in _call_driver_operation
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     driver_method(context, db_entity)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/opt/stack/neutron-lbaas/neutron_lbaas/drivers/a10networks/driver_v2.py", line 74, in create
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     self.driver.a10.pool.create(context, pool)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/a10-neutron-lbaas/a10_neutron_lbaas/v2/handler_pool.py", line 54, in create
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     c, context, pool)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/a10-neutron-lbaas/a10_neutron_lbaas/v2/handler_pool.py", line 36, in _set
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     axapi_args=args)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/slb/service_group.py", line 63, in create
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     **kwargs)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/slb/service_group.py", line 59, in _set
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     self._post(action, params, **kwargs)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/base.py", line 62, in _post
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     return self._request('POST', action, params, **kwargs)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/base.py", line 38, in _request
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     **kwargs)
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/axapi_http.py", line 230, in request
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     r, action=extract_method(api_url))
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin   File "/usr/local/src/acos-client/acos_client/v21/responses.py", line 169, in raise_axapi_ex
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin     raise ae.ACOSException(code, response['response']['err']['msg'])
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin ACOSException: 1162 Invalid integer. Parameter service_group.lb_method in method slb.service_group.create. The valid scope is 0 - 17.
2017-02-08 16:02:04.367 TRACE neutron_lbaas.services.loadbalancer.plugin
2017-02-08 16:02:04.518 ERROR neutron.api.v2.resource [req-6885af32-bfe7-4e5a-925f-f05a780d8338 admin e0142dcdf4904e05a40c2675db637473] create failed
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource Traceback (most recent call last):
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron/neutron/api/v2/resource.py", line 83, in resource
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     result = method(request=request, **args)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron/neutron/api/v2/base.py", line 410, in create
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     return self._create(request, body, **kwargs)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/usr/local/lib/python2.7/dist-packages/oslo_db/api.py", line 147, in wrapper
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     ectxt.value = e.inner_exc
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/usr/local/lib/python2.7/dist-packages/oslo_utils/excutils.py", line 204, in __exit__
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     six.reraise(self.type_, self.value, self.tb)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/usr/local/lib/python2.7/dist-packages/oslo_db/api.py", line 137, in wrapper
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     return f(*args, **kwargs)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron/neutron/api/v2/base.py", line 521, in _create
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     obj = do_create(body)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron/neutron/api/v2/base.py", line 503, in do_create
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     request.context, reservation.reservation_id)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/usr/local/lib/python2.7/dist-packages/oslo_utils/excutils.py", line 204, in __exit__
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     six.reraise(self.type_, self.value, self.tb)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron/neutron/api/v2/base.py", line 496, in do_create
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     return obj_creator(request.context, **kwargs)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron-lbaas/neutron_lbaas/services/loadbalancer/plugin.py", line 753, in create_pool
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     self._call_driver_operation(context, driver.pool.create, db_pool)
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource   File "/opt/stack/neutron-lbaas/neutron_lbaas/services/loadbalancer/plugin.py", line 479, in _call_driver_operation
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource     raise loadbalancerv2.DriverError()
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource DriverError: An error happened in the driver
2017-02-08 16:02:04.518 TRACE neutron.api.v2.resource
```

And the following running config

```
slb template persist source-ip 12efc4aa-f021-4441-aaef-40f15e96cda4
!
!
slb virtual-server a6f4c17d-351c-4f0a-a364-7f896a7b421d 10.90.102.187
   port 80  http
      name 6df625e0-6fb7-4c1c-b109-81e9412c428a
```

**Comment by Cedev** [comment]
It looks like that's an acos-client bug.

**Comment by Cedev** [comment]
I created a SOURCE_IP persistence pool successfully on a 4.1.0 api vthunder.

```
slb template persist source-ip 35fa24ca-0c2a-4804-a73c-6d321c323373
!
slb service-group 35fa24ca-0c2a-4804-a73c-6d321c323373 tcp
  method src-ip-hash
!
slb virtual-server f79ef1d3-afb2-45c4-bc50-5e728f6832fa 10.90.102.189
  port 80 http
    name 20c39925-35a3-4f81-842c-743bae1c6a26
    extended-stats
    service-group 35fa24ca-0c2a-4804-a73c-6d321c323373
```

**Pull request merged**